lib/Optimizer/Transforms/DecompositionPatterns.cpp

```diff
@@ -0,0 +1,871 @@
+    Value halfAngle = createDivF(loc, angle, 2.0, rewriter);
+    Value negHalfAngle = rewriter.create<arith::NegFOp>(loc, halfAngle);
+
+    rewriter.create<quake::RzOp>(loc, halfAngle, noControls, target);
+    rewriter.create<quake::XOp>(loc, control, target);
+    rewriter.create<quake::RzOp>(loc, negHalfAngle, noControls, target);
+    rewriter.create<quake::XOp>(loc, control, target);
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+// quake.rz(θ) target
+// ──────────────────────────────────
+// quake.phased_rx(π/2, 0) target
+// quake.phased_rx(-θ, π/2) target
+// quake.phased_rx(-π/2, 0) target
+struct RzToPhasedRx : public OpRewritePattern<quake::RzOp> {
+  using OpRewritePattern<quake::RzOp>::OpRewritePattern;
+
+  void initialize() { setDebugName("RzToPhasedRx"); }
+
+  LogicalResult matchAndRewrite(quake::RzOp op,
+                                PatternRewriter &rewriter) const override {
+    if (!op.getControls().empty())
+      return failure();
+    if (!quake::isAllReferences(op))
+      return failure();
+
+    // Op info
+    Location loc = op->getLoc();
+    Value target = op.getTarget();
+    Value angle = op.getParameter();
+    if (op.isAdj())
+      angle = rewriter.create<arith::NegFOp>(loc, angle);
+    Type angleType = op.getParameter().getType();
+
+    // Necessary/Helpful constants
+    ValueRange noControls;
+    Value zero = createConstant(loc, 0.0, angleType, rewriter);
+    Value pi_2 = createConstant(loc, M_PI_2, angleType, rewriter);
+    Value negPi_2 = rewriter.create<arith::NegFOp>(loc, pi_2);
+    Value negAngle = rewriter.create<arith::NegFOp>(loc, angle);
+
+    std::array<Value, 2> parameters = {pi_2, zero};
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+    parameters[0] = negAngle;
+    parameters[1] = pi_2;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+    parameters[0] = negPi_2;
+    parameters[1] = zero;
+    rewriter.create<quake::PhasedRxOp>(loc, parameters, noControls, target);
+
+    rewriter.eraseOp(op);
+    return success();
+  }
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// Populating pattern sets
+//===----------------------------------------------------------------------===//
+
+void cudaq::populateWithAllDecompositionPatterns(RewritePatternSet &patterns) {
+  // clang-format off
+  patterns.insert<
+    // HOp partterns
+    HToPhasedRx,
+    // SOp partterns
+    SToPhasedRx,
+    // TOp partterns
+    TToPhasedRx,
+    // XOp partterns
+    CXToCZ,
+    CCXToCCZ,
+    XToPhasedRx,
+    // YOp partterns
+    YToPhasedRx,
+    // ZOp partterns
+    CZToCX,
+    CCZToCX,
+    ZToPhasedRx,
+    // R1Op partterns
+    CR1ToCX,
+    R1ToPhasedRx,
+    // RxOp partterns
+    CRxToCX,
+    RxToPhasedRx,
+    // RyOp partterns
+    CRyToCX,
+    RyToPhasedRx,
+    // RzOp partterns
+    CRzToCX,
+    RzToPhasedRx
+  >(patterns.getContext());
```

Comment:
These patterns have a lot of boilerplate code. Can't we use tablegen and/or C++ templates to eliminate it?

Reply:
The problem is that some of these patterns can get a bit more involved than something that would be easy to implement using tablegen, e.g., when handling adjoints. Also, the native code does give us more power to define patterns names and, eventually, additional tags. 

I'm also entertaining the idea to specialize the patterns' benefit depending on the target basis, e.g., when converting to the `phased_rx, cz` basis, this pass should prioritize patterns which output those gates. (I'm not sure we would have such level of control with patterns defined using tablegen.)

You are right that some can benefit from C++ templates or relying on free functions. I will look into it.